src/memory/mmap_bytebuf.rs

```diff
@@ -5,181 +5,125 @@
+    const ALLOCATION_SIZE: usize = u32::MAX as usize;
+
+    /// Creates a new byte buffer with the given initial length.
     pub fn new(len: usize) -> Result<Self, String> {
-        let mmap = if len == 0 {
-            None
-        } else {
-            Some(Mmap::new(len)?)
-        };
-        Ok(Self { mmap })
+        if len > isize::max_value() as usize {
+            return Err("`len` should not exceed `isize::max_value()`".into());
+        }
+        let mem = VirtualMemory::new(Self::ALLOCATION_SIZE)?;
+        Ok(Self { mem, len })
     }
 
+    /// Reallocates the virtual memory with the new length in bytes.
     pub fn realloc(&mut self, new_len: usize) -> Result<(), String> {
-        let new_mmap = if new_len == 0 {
-            None
-        } else {
-            let mut new_mmap = Mmap::new(new_len)?;
-            if let Some(cur_mmap) = self.mmap.take() {
-                let src = cur_mmap.as_slice();
-                let dst = new_mmap.as_slice_mut();
-                let amount = src.len().min(dst.len());
-                dst[..amount].copy_from_slice(&src[..amount]);
-            }
-            Some(new_mmap)
-        };
-
-        self.mmap = new_mmap;
+        // This operation is only actually needed in order to make the
+        // Vec-based implementation less inefficient. In the case of a
+        // virtual memory with preallocated 4GB of virtual memory pages
+        // we only need to adjust the `len` field.
+        self.len = new_len;
         Ok(())
     }
 
+    /// Returns the current length of the virtual memory.
+    #[inline]
     pub fn len(&self) -> usize {
-        self.mmap.as_ref().map(|m| m.len).unwrap_or(0)
+        self.len
     }
 
+    /// Returns a shared slice over the bytes of the virtual memory allocation.
+    #[inline]
     pub fn as_slice(&self) -> &[u8] {
-        self.mmap.as_ref().map(|m| m.as_slice()).unwrap_or(&[])
+        &self.mem.as_slice()[..self.len]
     }
 
+    /// Returns an exclusive slice over the bytes of the virtual memory allocation.
+    #[inline]
     pub fn as_slice_mut(&mut self) -> &mut [u8] {
-        self.mmap
-            .as_mut()
-            .map(|m| m.as_slice_mut())
-            .unwrap_or(&mut [])
+        &mut self.mem.as_slice_mut()[..self.len]
     }
 
+    /// Writes zero to the used bits of the virtual memory.
+    ///
+    /// # Note
+    ///
+    /// If possible this API should not exist.
     pub fn erase(&mut self) -> Result<(), String> {
-        let len = self.len();
-        if len > 0 {
-            // The order is important.
-            //
-            // 1. First we clear, and thus drop, the current mmap if any.
-            // 2. And then we create a new one.
-            //
-            // Otherwise we double the peak memory consumption.
-            self.mmap = None;
-            self.mmap = Some(Mmap::new(len)?);
-        }
+        self.mem = VirtualMemory::new(Self::ALLOCATION_SIZE)?;
```

Comment:
Is there a test that verifies that the a new mapping is zeroed? I am asking because I am not so sure that this is guaranteed on windows.

Reply:
I think it is safe to assume that even on windows it will be initialized to zeros:
https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualalloc

Reply:
Okay.